app/assets/javascripts/labels.coffee

```diff
@@ -1,35 +1,82 @@
 ready = ->
   $('.my-tag').select2
+    width: '100%'
     tags: true
     multiple: true
     minimumInputLength: 1
+    maximumSelectionLength: 5
+    closeOnSelect: false
+
+    ajax: {
+      url: '/labels/search',
+      dataType: 'json',
+      data: (params) ->
+        query = {
+          search: params.term,
+          page: params.page || 1
+        }
+        return query;
+    }
+
+    ajax: {
+      url: '/labels/search',
+      dataType: 'json',
+      data: (params) ->
+        query = {
+          search: params.term,
+          page: params.page || 1
+        }
+        return query;
+    }
+
+    templateSelection: (data, container) ->
+      text = data.text;
+      bg_color = "#" + ($(data.element).attr("label_color") || data.label_color || "e4e4e4");
+      color = "#" + ($(data.element).attr("label_font_color") || data.label_font_color || "000000");
+
+      if data.newTag == true
+        text += " (new)";
+
+      $(container).css({"background-color" : bg_color, "color" : color});
+      $(container).children("span").css("color", color);
+
+      $template = $('<span></span>').text(text);
+      $template.css({"font-size" : "80%", "font-weight" : "bold"});
+      return $template;
+
+    templateResult: (data) ->
+      if (data.loading)
+        return data.text;
+
+      text = data.text;
+      bg_color = "#" + (data.label_color || "e4e4e4");
+      color = "#" + (data.label_font_color || "000000");
+
+      if data.newTag == true
+        text += " (new)";
+
+      $template = $('<div></div>').text(text).addClass("task_label");
+      $template.css({"background-color" : bg_color, "color" : color});
+      return $template;
+
     createTag: (params) ->
-      term = $.trim(params.term)
-      if term == ''
-        return null
-      {
-        id: term
-        text: term
+      if params.term == '' || params.term.length > 15
+        return null;
+
+      return {
+        id: params.term,
+        text: params.term,
         newTag: true
       }
-    width: '100%'
-    tokenSeparators: [
-      ','
-      ' '
-    ]
```

Comment:
The problem with having space as a tokenSeparator is that you cannot create labels like "Data Structures" anymore because it will separate on the space. I added the colon as a separator though.